lib/bolt/apply_target.rb

```diff
@@ -30,6 +30,10 @@ def self.from_asserted_args(uri = nil,
     end
     # rubocop:enable Lint/UnusedMethodArgument
 
+    def self._pcore_type
+      Target
```

Comment:
This should be `ApplyTarget`. It also seems like we don't need to define this if the datatype includes `Puppet::Pops::Types::PuppetObject`? Is there a reason to set both?

Reply:
I don't think this would be `ApplyTarget` since we don't define that as a Puppet type. The `Bolt::ApplyTarget` class is just an implementation class for the `Target` Puppet type.

`_pcore_type` is used by `ToDataConverter` here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/serialization/to_data_converter.rb#L247 . It's probably not strictly necessary to have this since apply targets are never serialized.

Reply:
It actually looks like this is already defined by the datatype creator? https://github.com/puppetlabs/puppet/blob/master/lib/puppet/datatypes.rb#L167 And...it seems like that overrides this local definition whether it's there or not? Which makes sense - I'm not sure where in this file `Target` would even be defined.

Reply:
But `Target` here is a reference to `Bolt::Target`, not to the Puppet Target datatype. Something seems fishy.

Reply:
@lucywyman That code you linked is really suspicious. It does look like it's meant to make exactly what we're doing here unnecessary. But it checks whether `implementation_class` already includes `PuppetObject`, and if so, it does some work that involves including `PuppetObject`. I wonder if the conditional is simply meant to be reversed?

Reply:
I don't think the problem is only that the logic is backward... The alternate case calls `register_implementation` which looks to be the way that Puppet associates a Ruby type with a Puppet type if the Ruby type _doesn't_ have `_pcore_type`. In that case, it may just be that the `include PuppetObject` that Puppet does is redundant (since it's already guaranteed the class descends from `PuppetObject`). But if so, that means we _only_ need to include the class and it will handle the proper definition of `_pcore_type` for us.

Reply:
The reason I think the method I linked is what ends up being used is that when I call `_pcore_type` on this class I get a `Puppet::Pops::Types::PObjectType`, which...seems unlikely to be coming from this call, but seems very likely to be coming from that method.

All of this to say: I agree with you Nick that I think we only need to include PuppetObject on the datatypes themselves, and don't need this definition (which...seems like it will only be confusing)